ISPN-15523 Fix the blocking publisher in the restore procedure

jabolina opened the pull request:
https://issues.redhat.com/browse/ISPN-15523

Just updating the flowables. See the comments: https://github.com/infinispan/infinispan/pull/12131#discussion_r1532260892

jabolina commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -211,32 +212,16 @@ public CacheBackupEntry next() {
+      });
+
+      return Flowable.fromPublisher(blockingManager.blockingPublisher(f))
+            .count()
```
I am not sure if this count here uses the correct threads.

wburns replied:
This is fine, the key is the BlockingManager#blockingPublisher method. 

```
    * Provided a publisher that is known to block when subscribed to. Thus if the thread that subscribes in a non
    * blocking thread we will instead subscribe on a blocking thread and observe on a non blocking thread for each
    * published value.
```
So any downstream operator will be ran on a non blocking thread for all published values.

wburns commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -193,10 +197,7 @@ public CacheBackupEntry next() {
-         Single<Long> restored = Flowable.fromIterable(() -> backupEntries)
+         return Flowable.fromIterable(() -> backupEntries)
                .rebatchRequests(batchSize)
                .map(entry -> {
```
This and rest of the operators here can be moved out of this publisher and added to the one below after it is wrapped by the BlockingManager. We only want the blocking code to hit the iterator, anything else should be outside of the using if possible.

wburns commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -122,56 +120,62 @@ public CompletionStage<Void> restore(ZipFile zip) {
+         SerializationContextRegistry ctxRegistry = SecurityActions.getGlobalComponentRegistry(cm).getComponent(SerializationContextRegistry.class);
+         ImmutableSerializationContext serCtx = ctxRegistry.getPersistenceCtx();
+
+         int batchSize = (int) (Long.highestOneBit(ProcessorInfo.availableProcessors()) << 1);
```
You might want to increase this batch size a bit, this isn't a parallelized operation in regards to the CPU. It will only be parallel when the put command is asynchronous and that is done via network most likely by updating backups or through a store async.

wburns replied:
I would suggest using the state transfer chunk size to allow tuning.

wburns commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -122,56 +120,61 @@ public CompletionStage<Void> restore(ZipFile zip) {
-         stages.dependsOn(cs);
-      }
-      return stages.freeze();
+      return CompletionStages.performConcurrently(resources, ProcessorInfo.availableProcessors(), blockingManager.asExecutor("cache-restore"), cacheName -> {
```
We should be using the non blocking executor here. No reason to use the blocking manager, it will actually make it worse due to how BlockingManager#blockingPublisher works.

wburns commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -193,50 +196,44 @@ public CacheBackupEntry next() {
+               cmd.setInternalMetadata(entry.internalMetadata);
+               return cmd;
+            })
+            .flatMap(cmd -> RxJavaInterop.voidCompletionStageToFlowable(invocationHelper.invokeAsync(cmd, 1)), batchSize)
```
Just a tip, but this internally has a parallelism to limit how many outstanding operations (in this case put commands) can be done.  The default is Flowable.bufferSize which is 128. I _think_ that should be fine for this case as this is not a CPU based parallelism anyways and is limited by network/disk instead.

wburns commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -122,56 +120,61 @@ public CompletionStage<Void> restore(ZipFile zip) {
-      }
-      return stages.freeze();
+      return CompletionStages.performConcurrently(resources, ProcessorInfo.availableProcessors(), blockingManager.asExecutor("cache-restore"), cacheName -> {
+         long start = System.currentTimeMillis();
```
Thinking more should we do it per cache or for all of them together? And actually I am thinking we should probably just change these to be TRACE/DEBUG messages stating start and end of each cache instead maybe.

jabolina replied:
A start/top message per cache seems good enough for debugging.

wburns commented on server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java:
```diff
@@ -193,50 +196,44 @@ public CacheBackupEntry next() {
+      boolean valueMarshalling = !scm.getValueStorageMediaType().isBinary();
+
+      int batchSize = SecurityActions.getCacheConfiguration(cache).clustering().stateTransfer().chunkSize();
+      return Flowable.fromPublisher(blockingManager.blockingPublisher(f))
```
This looks a lot cleaner now imo, great work!

infinispanrelease commented:
Image pushed for Jenkins build [#5](https://ci.infinispan.org/job/Infinispan/job/PR-12178/5/):
```
quay.io/infinispan-test/server:PR-12178
```

ryanemerson commented:
@jabolina This seems to fix the issue. I executed the test 20 times (once the build issue was resolved) and there was a single failure, which was not related to the restore procedure.

https://ci.infinispan.org/job/Infinispan%20Operator/job/PR-2036/

jabolina commented:
Thanks for the help, @ryanemerson! I'll remove the label so we can get this in, and later I'll add the commit to the open backport.

infinispanrelease commented:
Image pushed for Jenkins build [#6](https://ci.infinispan.org/job/Infinispan/job/PR-12178/6/):
```
quay.io/infinispan-test/server:PR-12178
```

wburns reviewed (approved):
LGTM, @ryanemerson can you confirm?

infinispanrelease commented:
Image pushed for Jenkins build [#7](https://ci.infinispan.org/job/Infinispan/job/PR-12178/7/):
```
quay.io/infinispan-test/server:PR-12178
```

ryanemerson commented:
@jabolina [org.infinispan.server.security.authorization.AuthorizationKerberosIT$Rest.testRestAdminsMustAccessBackupsAndRestores](https://ci.infinispan.org/job/Infinispan/job/PR-12178/7/testReport/junit/org.infinispan.server.security.authorization/AuthorizationKerberosIT$Rest/testRestAdminsMustAccessBackupsAndRestores/) looks related

jabolina commented:
@ryanemerson Good catch! Updated with a fix.

The issue was that we ran both HotRod and REST tests in the same container, so we had configuration from HotRod with indexing enabled "leaking" into the REST. The operations with the index enabled would block somewhere, causing the issue.

I've updated it to consider the "restore from file" blocking. Initially, I assumed that only reading the zip was blocking. Now, we have even the put operation considered blocking. This change will "hide" if an operation blocks by mistake.

infinispanrelease commented:
Image pushed for Jenkins build [#8](https://ci.infinispan.org/job/Infinispan/job/PR-12178/8/):
```
quay.io/infinispan-test/server:PR-12178
```

ryanemerson commented:
Thanks @jabolina
